packages/compass-import-export/src/modules/CursorExporter.ts

```diff
@@ -0,0 +1,69 @@
+import type { AbstractCursor } from 'mongodb';
+import { EventEmitter } from 'events';
+import * as stream from 'stream';
+import { promisify } from 'util';
+import { createCSVFormatter, createJSONFormatter } from '../utils/formatters';
+const pipeline = promisify(stream.pipeline).bind(stream);
+
+export interface CursorExporterOpts {
+  cursor: AbstractCursor;
+  type: 'csv' | 'json';
+  columns: Array<string> | boolean;
+  output: stream.Writable;
+  totalNumberOfDocuments?: number;
+}
+export class CursorExporter extends EventEmitter {
+  private _cursor: AbstractCursor;
+  private _output: stream.Writable;
+  private _formatter;
+  private _columns: Array<string> | boolean;
+  private _totalNumberOfDocuments: number;
+  private _exportedDocuments = 0;
+  constructor(opts: CursorExporterOpts) {
+    super();
+    this._cursor = opts.cursor;
+    this._formatter =
+      opts.type === 'csv' ? createCSVFormatter : createJSONFormatter;
+    this._output = opts.output;
+    this._columns = opts.columns ? opts.columns : true;
+    this._totalNumberOfDocuments = opts.totalNumberOfDocuments || 0;
+  }
+
+  async start(): Promise<void> {
+    await pipeline(
+      this._cursor.stream(),
+      this.getProgressTransformStream(),
+      this.getFormatter(),
+      this._output
+    );
+  }
+  private getFormatter() {
+    return this._formatter({ columns: this._columns });
+  }
+  private _getPercentageOfDocuments() {
+    if (this._totalNumberOfDocuments) {
+      return (this._exportedDocuments * 100) / this._totalNumberOfDocuments;
+    }
+    return 0;
```

Comment:
We don't always have the `totalNumberOfDocuments` (i.e. in aggregations) so I choose to return 0 instead of the percentage.

Reply:
Since we show the num of documents processed if the count is not available, why do we calculate the percentage here? Isn't the caller a better place to own this logic since it has all the knowledge? Would it be better to only `emit('processed')` here instead of trying to give a percentage back?

Reply:
> I choose to return 0 instead 

that makes sense but if there are 0 document to export that means that you are already at 100%. Just an edge case